torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -396,13 +399,80 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
     replica.contents = grad;
     // See Note [DDP Communication Hook]
     if (comm_hook_ == nullptr) {
-      replica.contents.div_(process_group_->getSize());
+      replica.contents.div_(divFactor_);
     }
     // The grad is modified in place and needs to be written back.
     return true;
   });
 }
 
+std::vector<std::vector<at::Tensor>> Reducer::getBucketTensors() const {
+ std::lock_guard<std::mutex> lock(mutex_);
+ std::vector<std::vector<at::Tensor>> bucketTensors;
+ for (const auto& bucket : buckets_) {
+  std::vector<at::Tensor> tensors;
+  tensors.reserve(bucket.replicas.size());
+  for (const auto& rep : bucket.replicas) {
+    tensors.push_back(rep.contents);
+  }
+  bucketTensors.push_back(std::move(tensors));
+ }
+ return bucketTensors;
+}
+
+bool Reducer::shouldRebuildBuckets() const {
+  // if find_unused_parameters_, we do not rebuild buckets.
+  return !find_unused_parameters_ && !has_rebuilt_bucket_;
+}
+
+void Reducer::setForwardPassWorkHandle(
+    std::shared_ptr<c10d::ProcessGroup::Work> forwardPassWorkHandle,
+    at::Tensor& tensor) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  // If there was a previous handle, it should have already completed. This is
+  // because we await this handle in the backward pass, and only reinstall in
+  // the next forward pass.
+  if (forwardPassWorkHandle_.workHandle) {
+    TORCH_INTERNAL_ASSERT(forwardPassWorkHandle_.workHandle->isCompleted());
```

Comment:
add an error message

Reply:
I think based on @pritamdamania87 's feedback here: https://github.com/pytorch/pytorch/pull/42577#discussion_r472613127, it would probably be best to remove this check